File: 백준/Gold/1700. 멀티탭 스케줄링/멀티탭 스케줄링.cc
#include <iostream>
#include <vector>
#include <unordered_map>
#include <set>

using namespace std;

int main() {
    int n, k; 
    cin >> n >> k; // 콘센트 개수, 사용 횟수

    vector<int> use(k); // 사용 순서 저장
    unordered_map<int, vector<int>> pos; // 전기기기별 사용 시점 저장
    set<int> hole; // 현재 콘센트에 꽂힌 기기 저장
    int result = 0;

    for (int i = 0; i < k; i++) {
        cin >> use[i];
        pos[use[i]].push_back(i); // 각 기기가 언제 사용되는지 저장
    }

    for (int i = 0; i < k; i++) {
        int now = use[i];
        pos[now].erase(pos[now].begin()); // 현재 사용한 시점 제거
        
        // 이미 꽂혀있다면 그냥 넘어감
        if (hole.find(now) != hole.end()) continue;

        // 콘센트에 자리가 남아있다면 꽂기
        if (hole.size() < n) {
            hole.insert(now);
        } else {
            // 앞으로 사용될 기기 중 가장 나중에 등장하는 것 제거
            int lastUse = -1, removeDevice = -1;
            for (int h : hole) {
                if (pos[h].empty()) { // 앞으로 안 쓰이는 기기가 있다면 바로 제거
                    removeDevice = h;
                    break;
                }
                // 가장 나중에 등장하는 기기 찾기
                if (pos[h].front() > lastUse) {
                    lastUse = pos[h].front();
                    removeDevice = h;
                }
            }

            // 콘센트에서 제거 후 새 기기 꽂기
            hole.erase(removeDevice);
            hole.insert(now);
            result++;
        }
    }

    cout << result;
}

File: 백준/Gold/1700. 멀티탭 스케줄링/멀티탭 스케줄링.py
from collections import defaultdict

def min_plug_removals(n, k, use):
    pos = defaultdict(list)  # 각 기기의 사용 시점을 저장할 딕셔너리
    hole = set()  # 현재 콘센트에 꽂힌 기기 저장
    result = 0  # 플러그를 뽑은 횟수

    # 각 기기가 언제 사용되는지 저장
    for i in range(k):
        pos[use[i]].append(i)

    # 차례대로 사용하기
    for i in range(k):
        now = use[i]  # 현재 사용해야 할 기기

        # 현재 사용한 시점 제거
        pos[now].pop(0)

        # 이미 꽂혀있다면 넘어가기
        if now in hole:
            continue

        # 구멍에 여유 있다면 꽂기
        if len(hole) < n:
            hole.add(now)
            continue  # 꽂았으면 다음으로 넘어감

        # 구멍에 여유 없다면 기존의 것 빼고 넣기
        last_use = -1
        remove_device = -1

        for h in hole:
            # 앞으로 사용되지 않을 기기 찾기 (즉시 제거)
            if not pos[h]:
                remove_device = h
                break

            # 가장 나중에 등장하는 기기 찾기
            if pos[h][0] > last_use:
                last_use = pos[h][0]
                remove_device = h

        # 콘센트에서 제거 후 새 기기 꽂기
        hole.remove(remove_device)
        hole.add(now)
        result += 1

    return result


# 입력 예제
n, k = map(int, input().split())  # 콘센트 개수, 사용 횟수
use = list(map(int, input().split()))  # 사용 순서

# 결과 출력
print(min_plug_removals(n, k, use))
